Remove ActiveSupport dependency

PericlesTheo opened the pull request.

PericlesTheo commented:
@nickcampbell18 if I can have some early feedback around this. A couple of questions:

1) I have 2 annoying Rubocop issues and I'm not sure how strict it is. Can you have a look and let me know? Speficially the one around testing with spies. I think it's a bit difficult to write the expectation with a spy and `have_received`.
2) https://github.com/gocardless/companies-house-rest/pull/60/files#diff-6114d6840cfa811f95551b1d8ec3bbb2L189 I've removed this check basically. I think I can add it back if I stub `SecureRandom`. I will try in a bit.
3) I have added active support in the Gemfile so I can the constant included in the specs. It's not really used for anything. Now, without it the spec in which I require `Instrumentation::ActiveSupport` fails which is the correct behaviour. Are you happy with that level of testing?

nickcampbell18 commented on README.md:
```diff
@@ -69,10 +69,15 @@ client = CompaniesHouse::Client.new(config)
+| -----------        | ----------- |
+| `:api_key`         | Required. The API key received after registration. |
+| `:endpoint`        | Optional. Specifies the base URI for the API (e.g. if using a self-hosted version) |
+| `:instrumentation` | Optional. Instruments the request/response (see Instrumentation for details) |
```
There's a way we can ship this change without making a breaking API change (for users depending on notifications). How about this?

1. If `defined?(ActiveSupport)`, we automatically configure `:instrumentation`. This is consistent with other Rails gems that essentially autoconfigure if in a Rails environment ([example](https://github.com/yammer/circuitbox/blob/0c508f50f56531eb191438ed19c8d527df9a8e91/lib/circuitbox/configuration.rb#L26-L30)).
2. If you pass any other argument to `:instrumentation`, that is configured instead.

PericlesTheo replied:
lol perfect, `circuitbox` is using pretty much the same thing 😄 Will do that 🙏

PericlesTheo replied:
Do you prefer the name `Notifier` or `Instrumentation`?

nickcampbell18 replied:
No real preference here, as long as it's documented!

nickcampbell18 commented on README.md:
```diff
@@ -69,10 +69,15 @@ client = CompaniesHouse::Client.new(config)
+
+## Instrumentation
+By default, no instrumentation is being applied.
+If you are using Rails or the `ActiveSupport` gem, you can pass to the client `instrumentation: Intstrumentation::ActiveSupport`
```
`s/Intstrumentation/Instrumentation`

nickcampbell18 commented on Gemfile:
```diff
@@ -4,3 +4,5 @@ source 'https://rubygems.org'
 # See companies-house.gemspec
 gemspec
+
+gem "activesupport", require: false
```
Can we accomplish this with `add_development_dependency` in the gemspec?

PericlesTheo replied:
Probably yes, will do that.

PericlesTheo replied:
Actually, I just removed this completely it's not needed.

nickcampbell18 commented on spec/companies_house/client_spec.rb:
```diff
@@ -160,36 +174,39 @@
         expect(response).to eq(%w[item1 item2])
       end
 
+      # rubocop:disable RSpec/MultipleExpectations
```
This is fine - splitting this into separate tests would be rubbish, I don't think we should mind disabling the cop here.

PericlesTheo commented:
Addressed the comments but Rubocop is failing because it doesn't like the `have_received` business. I will try and see if there is a different way of doing it.

PericlesTheo commented:
👋 @nickcampbell18, thoughts?

nickcampbell18 commented on lib/companies_house/client.rb:
```diff
@@ -19,6 +22,7 @@ def initialize(config)
       @endpoint = URI(config[:endpoint] || ENDPOINT)
       @open_timeout = config[:open_timeout] || 60
       @read_timeout = config[:read_timeout] || 60
+      @instrumentation = default_instrumentation
```
Do you want to allow people to pass an instrumenter via `:instrumentation` in `config`? I don't really mind since it looks like it can be overridden later but it might make for a more consistent API?

```ruby
client = CompaniesHouse::Client.new(...)
client.instrumentation = ...
```

vs

```ruby
client = CompaniesHouse::Client.new(instrumentation: ...)
```

PericlesTheo replied:
fixed that 👍

PericlesTheo commented:
@nickcampbell18 Great! Is there anything else I need to do on my side?

PericlesTheo commented:
This is embarrassing... I actually never removed the dependency 😢 https://github.com/gocardless/companies-house-rest/blob/master/companies-house-rest.gemspec#L25.

I've removed it locally just now, run specs and rubocop and everything passed. Shall I create another PR or are you happy pushing it yourself @nickcampbell18

nickcampbell18 commented:
Done. v0.4.6 released.
